File: src/rust/src/physics_setup.rs
//! Jolt scaffolding: layer constants + register/factory lifecycle. The broad-phase
//! layer interface and filters live inside the joltc C wrapper.

use crate::jolt::{self, ObjectLayer};

pub struct PhysicsLayers;
impl PhysicsLayers {
    pub const NON_MOVING: ObjectLayer = 0;
    pub const MOVING: ObjectLayer = 1;
    pub const NUM_LAYERS: ObjectLayer = 2;
}

pub const JOLT_MAX_PHYSICS_JOBS: i32 = 1024;
pub const JOLT_MAX_PHYSICS_BARRIERS: i32 = 8;

pub fn register_jolt_callbacks() {
    unsafe { jolt::jph_register_callbacks() };
}

/// RAII factory lifecycle (jph_factory_create / destroy).
pub struct JoltScope;
impl JoltScope {
    pub fn new() -> Self {
        unsafe { jolt::jph_factory_create() };
        JoltScope
    }

    /// Leak the factory for the program lifetime: it must outlive every physics object.
    pub fn leak() {
        std::mem::forget(JoltScope::new());
    }
}
impl Drop for JoltScope {
    fn drop(&mut self) {
        unsafe { jolt::jph_factory_destroy() };
    }
}

File: src/zig/physics_setup.zig
// physics_setup — Jolt scaffolding: layer constants + register/Factory
// lifecycle. The broad-phase layer interface/filters live inside the joltc
// wrapper (2-layer NON_MOVING/MOVING scheme that collides everything).

const jolt = @import("jolt.zig");

pub const PhysicsLayers = struct {
    pub const NON_MOVING: jolt.ObjectLayer = 0;
    pub const MOVING: jolt.ObjectLayer = 1;
    pub const NUM_LAYERS: jolt.ObjectLayer = 2;
};

pub const JOLT_MAX_PHYSICS_JOBS: i32 = 1024;
pub const JOLT_MAX_PHYSICS_BARRIERS: i32 = 8;

pub fn registerJoltCallbacks() void {
    jolt.jph_register_callbacks();
}

pub const JoltScope = struct {
    pub fn init() JoltScope {
        jolt.jph_factory_create();
        return .{};
    }
    pub fn deinit(_: *JoltScope) void {
        jolt.jph_factory_destroy();
    }
};
